lib-ee/emqx_ee_bridge/src/kafka/emqx_bridge_impl_kafka_consumer.erl

```diff
@@ -0,0 +1,378 @@
+                reason => Reason2
+            }),
+            stop_client(ClientID),
+            throw(failed_to_start_kafka_consumer)
+    end.
+
+-spec on_stop(manager_id(), state()) -> ok.
+on_stop(_InstanceID, State) ->
+    #{
+        subscriber_id := SubscriberId,
+        kafka_client_id := ClientID
+    } = State,
+    stop_subscriber(SubscriberId),
+    stop_client(ClientID),
+    ok.
+
+-spec on_get_status(manager_id(), state()) -> connected | disconnected.
+on_get_status(_InstanceID, State) ->
+    #{
+        subscriber_id := SubscriberId,
+        kafka_client_id := ClientID,
+        kafka_topic := KafkaTopic
+    } = State,
+    case brod:get_partitions_count(ClientID, KafkaTopic) of
+        {ok, NPartitions} ->
+            do_get_status(ClientID, KafkaTopic, SubscriberId, NPartitions);
+        _ ->
+            disconnected
+    end.
+
+%%-------------------------------------------------------------------------------------
+%% `brod_group_subscriber' API
+%%-------------------------------------------------------------------------------------
+
+-spec init(_, consumer_state()) -> {ok, consumer_state()}.
+init(_GroupData, State) ->
+    ?tp(kafka_consumer_subscriber_init, #{group_data => _GroupData, state => State}),
+    {ok, State}.
+
+-spec handle_message(#kafka_message{}, consumer_state()) -> {ok, commit, consumer_state()}.
+handle_message(Message, State) ->
+    ?tp_span(
+        kafka_consumer_handle_message,
+        #{message => Message, state => State},
+        begin
+            #{
+                resource_id := ResourceId,
+                hookpoint := Hookpoint,
+                kafka_topic := KafkaTopic,
+                mqtt := #{
+                    topic := MQTTTopic,
+                    payload := MQTTPayload,
+                    qos := MQTTQoS
+                }
+            } = State,
+            FullMessage = #{
+                offset => Message#kafka_message.offset,
+                key => Message#kafka_message.key,
+                value => Message#kafka_message.value,
+                ts => Message#kafka_message.ts,
+                ts_type => Message#kafka_message.ts_type,
+                headers => maps:from_list(Message#kafka_message.headers),
+                topic => KafkaTopic
+            },
+            Payload =
+                case MQTTPayload of
+                    full_message ->
+                        FullMessage;
+                    message_value ->
+                        Message#kafka_message.value
+                end,
+            EncodedPayload = emqx_json:encode(Payload),
+            MQTTMessage = emqx_message:make(ResourceId, MQTTQoS, MQTTTopic, EncodedPayload),
+            _ = emqx:publish(MQTTMessage),
+            emqx:run_hook(Hookpoint, [FullMessage]),
+            emqx_resource_metrics:received_inc(ResourceId),
+            %% note: just `ack' does not commit the offset to the
+            %% kafka consumer group.
+            {ok, commit, State}
```

Comment:
I'll that parameter as a configuration.

But just for my own understanding: what do you mean by duplicates?  The chance of having successfully processed a few Kafka messages and then the worker dying before commiting the offset?

I think the ack/commit is asynchronous in this case because of this sequence: [1](https://github.com/kafka4beam/brod/blob/ef97325545eed0ae406733bf177090b7a6a192b3/src/brod_group_subscriber_worker.erl#L80) -> [2](https://github.com/kafka4beam/brod/blob/ef97325545eed0ae406733bf177090b7a6a192b3/src/brod_topic_subscriber.erl#L497) -> [3](https://github.com/kafka4beam/brod/blob/ef97325545eed0ae406733bf177090b7a6a192b3/src/brod_topic_subscriber.erl#L521) -> [4](https://github.com/kafka4beam/brod/blob/ef97325545eed0ae406733bf177090b7a6a192b3/src/brod_consumer.erl#L269) .

But I'm not absolutely sure.  @zmstone , do you know if it's indeed async?

Reply:
sorry for the late reply.
yes, the offset commits are async.